File: January/Maximum Profit in Job Scheduling.java
class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int len = startTime.length;
        int[][] dp = new int[len][3];
        for(int i=0; i<len; i++){
            dp[i] = new int[]{startTime[i], endTime[i], profit[i]};
        }

        Arrays.sort(dp, (a,b)->a[1]-b[1]);
        TreeMap<Integer, Integer> map = new TreeMap<>();
        map.put(0, 0);
        for(var i:dp){
            int cur = map.floorEntry(i[0]).getValue() + i[2];
            if(cur > map.lastEntry().getValue()){
                map.put(i[1], cur);
            }
        }

        return map.lastEntry().getValue();
    }
}

File: January/cpp/Maximum Profit in job Scheduling.cpp
class Solution {
public:
    int jobScheduling(std::vector<int>& startTime, std::vector<int>& endTime, std::vector<int>& profit) {
        int len = startTime.size();
        vector<vector<int>> dp(len,vector<int>(3));

        for (int i = 0; i < len; i++) {
            dp[i] = {startTime[i], endTime[i], profit[i]};
        }

        sort(dp.begin(), dp.end(), [](const auto& a, const auto& b) {
            return a[1] < b[1];
        });

        std::map<int, int> map;
        map[0] = 0;

        for (const auto& i : dp) {
            int cur = (--map.lower_bound(i[0]))->second + i[2];
            if (cur > map.rbegin()->second) {
                map[i[1]] = cur;
            }
        }

        return map.rbegin()->second;
    }
};
